### ISSUE TITLE
How to emit default layout keycodes with modifier held

### ISSUE THREAD

**Issue by nazya** [rvaiya/keyd#488]
For example, my custom layout with unicode symbols generates the following with C-r pressed
leftcontrol down
cancel down
cancel up
0 down
0 up
q down
q up
9 down
9 up
leftcontrol up

But I want to make
leftcontrol down
r down
r up
leftcontrol up

And I also want such behavior for any modifier and any letter key.
It seems it is possible to redefine C, C-S, C-M, C-S-M, etc  layers for each letter key, but I believe it should be a better way.

**Comment by nsbgn** [comment]
To clarify: do you mean that you want to undo custom mappings for the duration of the activation of any modifier?

**Comment by nazya** [comment]
Yes, it would also be great. 
But I'd like to clarify more.
I created a custom layout where some letter keys are remapped (it is like dvorak layout). And I also have another layout with unicode symbols.
The problem is that shortcuts like С-<letter_key> do not work.
For example, if I have the unicode layout toggled I cannot create a new tab by hitting C-t, I need to toggle the custom layout first. But the other detail is that my <t> letter is also remapped in the custom layout.  So I don't want to undo all custom mapping, I want to undo only the unicode layout.
But the solution with undoing all custom mapping will also be  very appreciated.

**Comment by nsbgn** [comment]
Ah, got it. Good to know. If I understand correctly, your goal is to not have the Unicode symbols on your typing layer interfere with modifier layers.

As you've noted, you *can* already achieve what you want by explicitly redefining each letter key:

```
[main]
f = φ
...

[alt]
f = A-f
...

[control]
f = C-f
...

[control+alt]
f = C-A-f
...
```

In fact, I'm fairly sure you can shorten this to just the `control` and `alt` layers --- combinations will follow automatically, since other modifiers will stack on top of it. For that reason, I think this is an acceptable solution.

Nevertheless, I agree that it is still somewhat cumbersome. My first intuition for avoiding it would be to keep everything as is and just bind `leftcontrol` and friends to `swap(control)`, so that you deactivate your current layer while pressing control, like this:

```
[main]
leftcontrol = swap(control)
leftalt = swap(alt)
f = φ
...

[control]
leftalt = layer(alt)

[alt]
leftcontrol = layer(control)
```

However, this doesn't work, because `swap(control)` doesn't actually activate the `control` modifier. @rvaiya, shouldn't it?

**Comment by rvaiya** [comment]
> However, this doesn't work, because swap(control) doesn't actually activate the control modifier. @rvaiya, shouldn't it?

`main` is the default layout layer,  so you can't swap it out. That is, it is the layer consulted after all other layers have been exhausted (if swap worked, `control` would activate but the next key would be a noop since nothing would undergird the layer stack).

> In fact, I'm fairly sure you can shorten this to just the control and alt layers

This is the correct approach, so you would need something like:

```
[shift]
f = S-f
[meta]
f = M-f
[alt]
f = A-f
[control]
f = C-f
```

A bit cumbersome, but better than having to account for the power set of modifier combinations. 

It is important for the user to bear in mind that symbols like `φ` are just syntactic sugar for `macro(<$display_server compose sequence>)`, with all the attendant implications. I've considered special casing these symbols since it seems like a leaky abstraction, but there are good arguments against adding more magic (probably best discussed in a dedicated issue).

**Comment by rvaiya** [comment]
> ```
> [main]
> f1 = toggle(layer)
> 
> [layer]
> leftcontrol = swap(control)
> ```

I assume in this scenario the user keeps `layer` permanently activated, and defines `f = φ` within it (making it the de facto 'layout'). I suppose this approach could work, but it would require the user to explicitly toggle `layer` on initialization and would interfere with actions like `clear`. 

An earlier version of keyd actually had a dedicated feature to allow the definition of distinct modifier and letter layouts, but it was scrapped for a variety of reasons which I won't dwell on here.

The current recommended way to define a modifier set which deviates from the layout is to explicitly override each default modifier layer and rely on stacking semantics, as above.

**Comment by nsbgn** [comment]
> I suppose this approach could work, but ...

Yeah, the fact that you can't `swap` the main layer makes this alternative approach less than ideal. To be clear, I'm not advocating spending any effort to make it work (although I can make an argument for it) --- I was just observing that, while `layer(layer)` activates any modifiers associated with `layer`, `swap(layer)` does not, which is quite surprising even outside of this issue. If it is intended, perhaps it merits a note in the manual. Since it's otherwise tangential to this issue, that is all :P

**Comment by rvaiya** [comment]
Sorry, I was mistaken in my earlier response. `swap` doesn't work in this case since it replaces the toggled layer with another *toggled* layer, which is not what you want.

> I was just observing that, while layer(layer) activates any modifiers associated with layer, swap(layer) does not.

It does, it just replaces the toggled layer (`layer`), which is not what you want in this case.

**Comment by nazya** [comment]
Thanks!
